### PR TITLE
feat(efb): add settings page with audio options

### PR DESCRIPTION
**Pull request by Benjozork**
## Summary of Changes

- [x] Add PTU in cockpit option
- [x] Add Exterior Master Volume option
- [x] Add Engine Interior Volume option
- [x] Add Wind Interior Volume option
- [x] Persist settings between flights
- [ ] Add numerical indicator next to slider

Discord username (if different from GitHub): someperson#4953

## Testing instructions

* Make sure all four options work in real-time.
* Make sure all four options persist after MSFS restarts (DO NOT use ALT+F4, close MSFS through the main menu)

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Review by beheh — APPROVED** [review]

File: src/instruments/src/Common/simVars.tsx [hunk 1]
```diff
@@ -59,7 +77,7 @@ const context = React.createContext<{
 });
 const { Provider: InternalProvider }  = context;
 
-type UnitName = string | any; // once typings is next to tsconfig.json, use those units
+export type UnitName = string | any; // once typings is next to tsconfig.json, use those units
```
Comment (beheh) [inline src/instruments/src/Common/simVars.tsx:80]: Maybe let's not do this for now, this is more an internal placeholder at this time.

Comment (Benjozork) [reply src/instruments/src/Common/simVars.tsx]: changed to `string`

**Comment by Recka50** [comment]
Category: QA Tester/Trial
Name: Bepis#6926
Date of testing: Current date 20/02/2021
Version of the sim: 1.13.16.0
PR Tested: #3656
Tier of testing: 1
Changes to observe: Sound options for PTU barking in cockpit, interior engine sound volume, exterior master volume and wing sounds inside the cockpit

Testing technique: Spawned in cold and dark and ran through setup adjusting various volume settings relevant to the features tested.

Steps you've made to test the Pull Request:
1. Spawned cold and dark
2. Set PTU sounds to on
3. While PTU sounds are happening, tested turn off and on to check if sounds cut out and back in
4. Adjusted volume and spooled engines up and down at various volume sounds
5. Tested exterior sounds in flight
6: Tested sounds in flight for wings and spoilers at various volume levels

Flight notes (When done a full flight): N/A

Media: Insert screenshots, recordings, or anything relevant to the review here.
Video of testing methodology https://www.youtube.com/watch?v=1kuqrF0NGx4

Issues (If Any): No issues

Overall Rating: Good
Conclusions of the testing: The changes for the various volume settings work well in my testing, no issues found!

**Comment by stescouse** [comment]
Category: QA Tester
Name: stescouse#7271
Date of testing: 21/2/21
Version of the sim: 1.12.13.0
PR Tested: #3656
Tier of testing: 1
Changes to observe: Sound options for PTU

1. Spawned cold and dark at SBGL
2. Turned on PTU sounds in EFB.
3. Turned up volume for Exterior Master Volume and Engine Interior Volume.
4. Started APU and listened to both inside and outside the cockpit.
5. Started the APU Bleed and listened to inside of cockpit.
6. Started Engine 1 and listened to barking and engine runup inside and outside the cockpit.
7. Started Engine 2 and listened to engine runup inside and outside the cockpit.
8. Run both engines, and APU down and listened to sounds both inside and outside the cockpit at various volume levels.
9. Did full flight and listened to wind sounds at full volume and no volume.

Flight notes (When done a full flight): SBGL to SBGR.

Issues (If Any): No issues

Overall Rating: Good
Conclusions of the testing: Good variation of sound at different levels for Exterior Master Volume and Engine Interior Volume.  Wind volume increased sound in wind heard inside the cockpit.  Sounds added to realism and customisation. All performed well as outlined in the PR.

**Comment by marcman86** [comment]
Name: marcman86#4907
Date of testing: 21.2.2021
Version of the sim: 1.13.16.0
PR Tested: #3656
Tier of testing: 1
Changes to observe: EFB audio setups
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Turned on bat., ext. power and EFB, switch to settings page (Image1)
3.) Turned on PTU
4.) Started engines (first engine 1, than 2)
5.) Played with sliders
6.) Took off

Media:


Image1

- PTU back in cockpit
- Sliders effect soundvolume

Issues: 

- Wind slider don't effect air condition and bleed
- Engine slider don't effect APU
- Is middle position default (100)? Hard to get back to this point. Also its hard to know whats the value of the volume slider. Best would be a number box before the slider where you can type in the volume value (0-200) or a reset button.

Overall Rating: -

Conclusions of the testing: Needs improvements

Hope this helps :-)

**Comment by CAC14** [comment]
Category: QA Tester
Name: Cody Discord User #9027
Date of testing: Current date 21/02/2021
Version of the sim: Version the tester uses
PR Tested: #3656
Tier of testing: 1/2/3
Changes to observe: Check all the sounds and mess with them on tablet to see what happens

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Started at EDDM gate. Went into msfs setting and moved all audio to middle. Turned on external power.  Blower sounds started.  Brought all volume levels to min in EFB.  Still heard blower and switch sounds.    Switched on PTU in cockpit. 
2. Started apu.  No adjusting of settings in EFB changed apu or blower sounds. Inside sounds seem unaffected
3. Started engines.  Adjusting “engine interior volume” did change engine sounds.  Still no PTU sound after adjusting all volume levels.
4.  Taxied to run way and took off.  As wind volume increased adjusting it on EFB worked fine.  Engine volume was adjustable as well.  Exterior Master volume did lower sound when out of cockpit but even with it all the way down it did not mute it.  Switching PTU on and off still did nothing.
5.  Took video showing sound going on and off.  Also some of the controls seemed to stick or be hard to move.
6. Took a video of external master sound working but not muting sounds.
7. re tried starting engines and using ctrl+E.  Never heard PTU. (see 3rd video)  tried cranking and starting.  No luck.

Flight notes (When done a full flight):  PTU does not work.  others work as expected although External should go all the way to 0.  Maybe add internal sound slider for apu/blower.  Could apu be grouped in with engine sound?  (not currently)

Media: https://youtu.be/sXi0GcO0wzI https://youtu.be/oe81zXvKSKE https://youtu.be/gH-PrrFvZBY

Issues (If Any): PTU did not work at all.  Wind and Engine sounds worked as intended.  Exterior Master was able to lower volume but not mute it even when all the way down.  Apu was not affected when adjusting engine sound.

Overall Rating: To be improved.  
Conclusions of the testing: PTU does not work and Exterior master should have full range all the way down to mute.

**Comment by marcman86** [comment]
To listen to PTU check the PTU switch on the setup page to on.

Start **one** engine. When its AVAIL you will hear the PTU sound.

**Comment by Benjozork** [comment]
@CAC14 currently there is no "mute" option available. further sliders *may* be added in another PR but this is unlikely.

**Comment by CAC14** [comment]
> 
> 
> @CAC14 currently there is no "mute" option available. further sliders _may_ be added in another PR but this is unlikely.

i just meant moving the slider to 0 didn't bring the sound all the way down

**Comment by CAC14** [comment]
> 
> 
> To listen to PTU check the PTU switch on the setup page to on.
> 
> Start **one** engine. When its AVAIL you will hear the PTU sound.

I heard a tiny sound that sounds like a drill when stuck.  it was bairly load enough to hear.  i could bairly hear it over the apu noise.

**Comment by hotshotp** [comment]
> > @CAC14 currently there is no "mute" option available. further sliders _may_ be added in another PR but this is unlikely.
> 
> i just meant moving the slider to 0 didn't bring the sound all the way down

Yes, that is correct. It is a 6 decibel difference when going from the middle down. As for the drill, that is correct. You also cannot hear the apu in the cockpit, so i have no idea what you are talking about

**Comment by Benjozork** [comment]
> > 
> > 
> > @CAC14 currently there is no "mute" option available. further sliders _may_ be added in another PR but this is unlikely.
> 
> i just meant moving the slider to 0 didn't bring the sound all the way down

Yes that's intended. We will not provide the option to mute it. You can use Windows Sound Mixer for that.

**Comment by CAC14** [comment]
> 
> 
> > > @CAC14 currently there is no "mute" option available. further sliders _may_ be added in another PR but this is unlikely.
> > 
> > 
> > i just meant moving the slider to 0 didn't bring the sound all the way down
> 
> Yes, that is correct. It is a 6 decibel difference when going from the middle down. As for the drill, that is correct. You also cannot hear the apu in the cockpit, so i have no idea what you are talking about - everything that you have commented on is intentional

You can hear the bleed air from the apu in the cockpit

**Comment by HugoTro** [comment]
Category: QA Tester
Name: Hugo #3307
Date of testing: Current date 23/02/2021
Version of the sim: 1.13.16.0
PR Tested: #3656 
Tier of testing: 1
Changes to hear: Sounds volume moving when sliding the sliders

Testing technique: Hear-at
Steps you've made to test the Pull Request:
1. Spawned in C&D
2. Powered up the aircraft
3. Tested the PTU sound option with different views (cockpit, windows) while starting the engine each time
4. Tested the exterior master volume after engines started
5. Took off and tested the wind in cockpit slider


Issues (If Any): None

Overall Rating: Good
Conclusions of the testing: During the PTU test, PTU is audible in cabin with "PTU audible in cockpit" option disabled, which is good. Overall the sliders work, LGTM 👍

**Comment by stescouse** [comment]
Category: QA Tester
Name: stescouse#7271
Date of testing: 23/2/21
Version of the sim: 1.12.13.0
PR Tested: #3656
Tier of testing: 1
Changes to observe: Sound options for PTU

1. Spawned cold and dark at SBGL
2. Turned on PTU sounds in EFB.
3. Turned up volume for Exterior Master Volume and Engine Interior Volume.
4. Started APU and listened to both inside and outside the cockpit.
5. Started the APU Bleed and listened to inside of cockpit.
6. Started Engine 1 and listened to barking and engine runup inside and outside the cockpit.
7. Started Engine 2 and listened to engine runup inside and outside the cockpit.
8. Run both engines, and APU down and listened to sounds both inside and outside the cockpit at various volume levels.
9. Did full flight and listened to wind sounds at full volume and no volume.
10. Restarted the sim and observed that set sound levels were retained and the same as previous flight.

First Flight.
gggggg

Second Flight.

gghhhhjj


Flight notes (When done a full flight): SBGL to SBGR.

Issues (If Any): No issues

Overall Rating: Good
Conclusions of the testing: Good variation of sound at different levels for Exterior Master Volume and Engine Interior Volume. Wind volume increased sound in wind heard inside the cockpit. Sounds added to realism and customisation.  Volume settings were retained after restarting the sim.   All performed well as outlined in the PR.

**Comment by NonstopFlying** [comment]
Category: QA Trial
Name: NonstopFlying (Henrik) #6771
Date of testing: 23/2/21
Version of the sim: 1.13.16.0
PR Tested: #3656
Tier of testing: 1
Changes to observe: Audio Options on the EFB

Spawned in at Faro
Selected the different Audios and used the slider
Turn on and off the PTU option 
Listend to the difference between audio volume from the outside
Listend to the inside audio options and how they are effected via the sliders

Testing technique: Listening

Issues: Not really an issue, if you press and hold the slider and move the curser out of the EFB and let go of the press, you can move the slider without pressing anywhere. (maybe somehow add like a border to the EFB)
Media to the issue:

https://user-images.githubusercontent.com/78429170/108859023-1716c800-75ed-11eb-942a-199ac0243ea7.mp4

Conclusion: Good works like it is supposed to.

**Review by kberg93 — APPROVED** [review]
Category: QA Tester
Name: Kberg#0666
Date of testing: Current date 02/23/2021
Version of the sim: 1.13.16.0
PR Tested: #3656 
Tier of testing: 1
Changes to observe:
add settings page with audio options

Testing technique: Adjusting volume and listening.

Steps you've made to test the Pull Request: 
1. Spawned on runway, adjusted sound and listened for changes interior and exterior.
2. Took off and climbed to 5000ft and adjusted wind volume and listened for changes.
3. Went back and landed. Stopped engines put on PTU to cockpit sound. Started and listened for PTU, the PTU was audible in cockpit. Then i turned off the option to confirm it isnt audible when off again.
4. Restarted sim to check that settings were stored wich they was. 

Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.

Issues (If Any): None.

Overall Rating: Good.

Conclusions of the testing: Worked like a charm, no issues and it remembers last settings after restarting sim 100%. SGTM.

**Comment by Benjozork** [comment]
@wpine215 cannot make the slider snap at the moment. native html feature is not supported by coherent. will look into something custom later, for now I will add an indicator to show the current value to help align it to 50.

**Pull request merged**